jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java

```diff
@@ -527,31 +514,51 @@ void parseAndFillForContent()
 
             assert !_requestBuffer.hasRemaining();
 
+            int filled;
+            boolean compact = true;
+
+            // If the application has retained the content chunks then we must not overwrite content.
             if (_requestBuffer.isRetained())
             {
-                // The application has retained the content chunks,
-                // reacquire the buffer to avoid overwriting the content.
-                releaseRequestBuffer();
-                ensureRequestBuffer();
+                // If there is sufficient space available, we can top up the buffer rather than allocate a new one
+                ByteBuffer backing = _requestBuffer.getByteBuffer();
+                if (BufferUtil.space(backing) >= getHttpConfiguration().getMinInputBufferSpace())
```

Comment:
This check should test for `minInputBufferSpace > 0 && space > minInputBufferSpace` like HTTP/2 does.